tls/s2n_config.c

```diff
@@ -530,11 +530,15 @@ int s2n_config_set_verification_ca_location(struct s2n_config *config, const cha
     return err_code;
 }
 
-static int s2n_config_add_cert_chain_and_key_impl(struct s2n_config *config, struct s2n_cert_chain_and_key *cert_key_pair)
+int s2n_config_add_cert_chain_and_key_impl(struct s2n_config *config, struct s2n_cert_chain_and_key *cert_key_pair)
 {
+    POSIX_ENSURE_REF(config);
     POSIX_ENSURE_REF(config->domain_name_to_cert_map);
     POSIX_ENSURE_REF(cert_key_pair);
 
+    /* Validate that the certificate type is allowed by the security policy. */
+    POSIX_GUARD_RESULT(s2n_security_policy_validate_certificate_chain(config->security_policy, cert_key_pair));
+
```

Comment:
Doesn't this introduce potential ordering issues when building configs?

* If I add my certificates before setting my security policy, those certificates are validated against the default security policy. So default policies can never enforce certificate restrictions.
* If I want to swap out my security policy later, and the old and new policies have different certificate restrictions, I may be stuck and have to just create an entirely new config.

In general we've tried to avoid checks like this because 1) You have to make sure you cover every entry point 2) Ordering of calls starts to matter. In the past, we've solved the problem by not validating the config until you try to set it on a connection. We don't have a config builder, but setting it on a connection essentially finalizes it.

Reply:
I removed the checks for `load_cert` and `s2n_config_set_cipher_preferences`, and pushed those validations to `s2n_connection_set_config`.

I left the check for `s2n_connection_set_cipher_preference` in because otherwise we'd be missing the case where
1. a config is set on the connection
2. a strict security policy override is set on the connection
But this does introduce an ordering dependency. Let me know if you'd prefer to remove it